File: apps/spell-circle-canvas/src/common/compose/reference/examples/foreground_verb.cpp
/** @file
 * foreground — a decoration painted over the children, which is what a
 * keyline that must survive the content it sits on needs.
 *
 * A reference example. It stands outside the sketch registry, so it is
 * photographed with `--frame` and never enters the plate sweep.
 */

#include <sigilcompose/brush/Decorations.h>
#include <sigilcompose/core/Core.h>
#include <sigilsketch/canvas/Sketch.h>

namespace sketch = sigil::sketch;

using namespace sigil::compose;

namespace {

constexpr SkSize kCanvas = {620, 250};
constexpr SkColor4f kGround = hexColor(0x14181d);
constexpr SkColor4f kPlate = hexColor(0x223039);
constexpr SkColor4f kInk = hexColor(0xe8eef2);
constexpr SkColor4f kKey = hexColor(0xf0f4f7);
constexpr SkColor4f kAsh = hexColor(0x8ea0ad);

/** A tile whose child fills the whole box, so a mark under the children
 *  would be covered and a mark over them is not. */
Element tile(const char* caption, Element plate) {
  return box()
      .column()
      .gap(10)
      .flexBasis(0)
      .flexGrow(1)
      .alignItems(Align::Center)
      .children(
          {plate.width(pct(100)).height(120).borderRadius({10}).children(
               {box()
                    .cover()
                    .borderRadius({10})
                    .fill(kPlate)
                    .justifyContent(Justify::Center)
                    .alignItems(Align::Center)
                    .children(
                        {text("A FULL-BLEED CHILD")
                             .font({.size = 14, .color = kInk, .track = 2})})}),
           text(caption).font({.size = 12, .color = kAsh})});
}

}  // namespace

struct ForegroundVerb {
  void setup(sketch::SketchContext& ctx) {
    ctx.canvas({.size = kCanvas, .background = kGround, .captureSeconds = 0});
    ctx.composer.render(describe());
  }

  Element describe() const {
    return box().row().gap(18).padding(24).children({
        // Under the fill and under the children: the child covers it.
        tile("background(keyline)",
             box().background(
                 stroke(3, Fill::color(kKey), PathFormat::Align::Inner))),
        // Over the children: the keyline stands.
        tile("foreground(keyline)",
             box().foreground(
                 stroke(3, Fill::color(kKey), PathFormat::Align::Inner))),
    });
  }
};

SIGIL_SKETCH(ForegroundVerb, "Reference · Compose",
             "a keyline under and over a full-bleed child")

File: apps/spell-circle-canvas/src/common/compose/reference/examples/foreground_verb.py
"""foreground — a decoration painted over the children, which is what a
keyline that must survive the content it sits on needs.

The Python twin of foreground_verb.cpp. A reference example: it is
rendered with ``sigil render`` and belongs to no sketch registry.
"""

from sigil import compose
from sigil.sketch import SketchContext, sketch

GROUND = "#14181d"
PLATE = "#223039"
INK = "#e8eef2"
KEY = "#f0f4f7"
ASH = "#8ea0ad"


def keyline() -> compose.PathFormat:
    return compose.stroke(3, KEY, compose.PathFormat.Align.Inner)


def tile(caption: str, plate: compose.Element) -> compose.Element:
    """A tile whose child fills the whole box, so a mark under the
    children would be covered and a mark over them is not."""
    return (
        compose.box(
            plate.width(compose.pct(100))
            .height(120)
            .corners(10)
            .children(
                compose.box(
                    compose.text("A FULL-BLEED CHILD", size=14, color=INK).fontTrack(2)
                )
                .cover()
                .corners(10)
                .fill(PLATE)
                .justify("center")
                .alignItems("center")
            ),
            compose.text(caption, size=12, color=ASH),
        )
        .column()
        .gap(10)
        .basis(0)
        .grow(1)
        .alignItems("center")
    )


@sketch(size=(620, 250), background=GROUND, capture_at=0)
class ForegroundVerb:
    def setup(self, ctx: SketchContext) -> None:
        ctx.render(self.describe())

    def describe(self) -> compose.Element:
        return (
            compose.box(
                # Under the fill and under the children: the child covers it.
                tile("background(keyline)", compose.box().background(keyline())),
                # Over the children: the keyline stands.
                tile("foreground(keyline)", compose.box().foreground(keyline())),
            )
            .row()
            .gap(18)
            .padding(24)
        )
